File: windows/src/Slipstream.Core/Discovery/GatewayProbeStrategy.cs
using System.Net;
using Slipstream.Core.Net;

namespace Slipstream.Core.Discovery;

/// <summary>
/// Spec §5 S2. Decisive for the phone-hotspot case, where the phone is the PC's
/// default gateway: one probe, no scanning, no multicast, cannot be defeated by
/// access-point behaviour.
/// </summary>
public sealed class GatewayProbeStrategy(PeerProbe probe) : IDiscoveryStrategy
{
    public string Name => "gateway-probe";

    public async Task<DiscoveredPeer?> FindAsync(LocalNetwork network, CancellationToken cancellationToken)
    {
        if (network.Gateway is null) return null;
        if (!LanGuard.IsLocal(network.Gateway)) return null;

        return await probe(new IPEndPoint(network.Gateway, SlipstreamPorts.Control), cancellationToken);
    }
}

File: android/core/src/main/kotlin/com/slipstream/core/discovery/GatewayProbeStrategy.kt
package com.slipstream.core.discovery

import com.slipstream.core.SlipstreamPorts
import com.slipstream.core.net.LocalNetwork
import java.net.InetSocketAddress

/**
 * S2: probe the network's default gateway on the control port. Spec §5: this is the
 * decisive strategy for the primary scenario — when the phone is the hotspot, it *is*
 * the PC's gateway, so the PC resolves the peer deterministically without scanning,
 * multicast, or any AP cooperation.
 */
class GatewayProbeStrategy(
    private val probe: PeerProbe,
) : DiscoveryStrategy {

    override val name: String = "gateway-probe"

    override suspend fun find(network: LocalNetwork): DiscoveredPeer? {
        val gateway = network.gateway ?: return null
        return probe.probe(InetSocketAddress(gateway, SlipstreamPorts.CONTROL))
    }
}
